### PR TITLE
[SYCL] Add range and nd_range CTAD support

### PR DESCRIPTION
**Pull request by rolandschulz**
Also replace one enable_if with static_assert for better diagnostic.

### CODE REVIEW

File: sycl/test/basic_tests/range_ctad.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include <CL/sycl.hpp>
+
+using namespace std;
+int main() {
+  cl::sycl::range one_dim_range(64);
+  cl::sycl::range two_dim_range(64, 1);
+  cl::sycl::range three_dim_range(64, 1, 2);
```
Comment (keryell) [inline sycl/test/basic_tests/range_ctad.cpp]: Interesting to combine new-style CTAD with old-style explicit constructor call...

**Comment by rolandschulz** [comment]
> Thanks for starting implementing my bigger plan. :-)
> [triSYCL/triSYCL#218](https://github.com/triSYCL/triSYCL/issues/218)

Are there any other we should add? I looked through all and couldn't find any other. 
The only other constructor I'm aware of where CTAD doesn't work is `accessor`. But for the `accessor` without `handler` argument it isn't obvious whether it should create a global placeholder accessor or a host accessor. And for the one with `handler` argument the `get_accessor` method seems a better choice.

**Comment by keryell** [comment]
> Are there any other we should add? I looked through all and couldn't find any other.
> The only other constructor I'm aware of where CTAD doesn't work is `accessor`. But for the `accessor` without `handler` argument it isn't obvious whether it should create a global placeholder accessor or a host accessor. And for the one with `handler` argument the `get_accessor` method seems a better choice.

Yes probably for the accessors `get_accessor()` is the way to go since we have `auto`.
The most obvious & useful case is inferring buffer type and dimensions with CTAD.

**Comment by romanovvlad** [comment]
@rolandschulz Please, rebase your changes.

**Pull request merged**